### PR TITLE
feat: Parcel card 2.0, client realm filters improvement

### PR DESCRIPTION
**Pull request by dudendy**

### CODE REVIEW

File: src/api/installations.api.js [hunk 1]
```diff
@@ -9,5 +9,39 @@ const installationsContract = ethersApi.makeContract(INSTALLATION_CONTRACT, INST
...
+                parcelId: ethersApi.formatBigNumber(item.parcelId),
+                installationId: ethersApi.formatBigNumber(item.installationId),
+            })));
+    },
+
+    async finalizeUpgrades(ids) {
+        const contractWithSigner = ethersApi.makeContractWithSigner(INSTALLATION_CONTRACT, INSTALLATIONS_ABI);
+        const transaction = await contractWithSigner.finalizeUpgrades(ids);
+
+        console.log('ids', ids);
```
Comment (dmutrokot) [inline src/api/installations.api.js:38]: remove please

Comment (dmutrokot) [reply src/api/installations.api.js:38]: also check other `console.log`s

File: src/components/ActiveListingButton/ActiveListingButton.js [hunk 2]
```diff
@@ -17,12 +17,10 @@ export default function ActiveListingButton({ item }) {
         let mounted = true;
 
         setListingLoading(true);
-        console.log(item)
 
         thegraphApi.getActiveListing(item.erc, item.id, item.type, item.category)
             .then(res => {
-                if(mounted) {
-                    console.log('res', res)
+                if(mounted && res) {
```
Comment (dmutrokot) [inline src/components/ActiveListingButton/ActiveListingButton.js:23]: space missed. And why would you need `res` condition? `listing` can be null

File: src/components/CopyToClipboard/CopyToClipboardBlock.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            followCursor
+            className={classes.tooltip}
+        >
+            <div
+                className={classNames(classes.block, className)}
+                onClick={(event) => copyText(event, text)}
+                onMouseEnter={() => setTooltipText(defaultTooltipText)}
+            >
+                <span>{children}</span>
+                {/* <ContentCopyIcon className={classes.blockIcon} /> */}
```
Comment (dmutrokot) [inline src/components/CopyToClipboard/CopyToClipboardBlock.js:42]: please remove

File: src/components/Items/Parcel/Parcel.js [hunk 4]
```diff
@@ -29,51 +36,82 @@ export default function Parcel({ parcel }) {
...
+                        parcel={parcel}
+                        imageSize={300}
+                        key={parcel.parcelId}
+                    />
+                </div>
 
-            <ParcelLink parcel={parcel} />
+                <div className={classes.boosts}>
+                    {Object.entries(boosts).map((boost, i) => {
+                        let key = boost[0];
```
Comment (dmutrokot) [inline src/components/Items/Parcel/Parcel.js:82]: please use `const`

File: src/components/Items/Parcel/ParcelChanneling.js [hunk 5]
```diff
@@ -34,6 +34,10 @@ export default function ParcelChanelling({ parcelId }) {
             }
         });
 
+        // realmApi.getParcelInfo(parcelId).then(res => {
```
Comment (dmutrokot) [inline src/components/Items/Parcel/ParcelChanneling.js:37]: please remove

File: src/components/Items/Parcel/ParcelChanneling.js [hunk 6]
```diff
@@ -82,6 +86,7 @@ export default function ParcelChanelling({ parcelId }) {
                                 <FudIcon className={classes.unactiveIcon} height={24} width={28} />
                             )
                         )}
+
```
Comment (dmutrokot) [inline src/components/Items/Parcel/ParcelChanneling.js:89]: seems that is no need for it

File: src/components/Items/Parcel/ParcelName.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import classNames from 'classnames';
```
Comment (dmutrokot) [inline src/components/Items/Parcel/ParcelName.js:2]: separate please

File: src/components/Items/ParcelImage/ParcelImage.js [hunk 8]
```diff
@@ -2,49 +2,74 @@ import React, { useEffect, useRef, useState } from 'react';
...
-        const context = canvas.getContext('2d');
+    const processColorsMap = (map) => {
+        let cache = [];
 
-        const drawRect = (width, height) => {
-            const parcelX = (parcelSize - width < parcel.coordinateX ? parcelSize - width : parcel.coordinateX) / 2;
-            const parcelY = (parcelSize - height < parcel.coordinateY ? parcelSize - height : parcel.coordinateY) / 2;
+        map.forEach((item) => {
+            if (!cache.length) {
+                cache[0] = []
```
Comment (dmutrokot) [inline src/components/Items/ParcelImage/ParcelImage.js:19]: semicolon missed, please also check formatting in other statements in `map`

File: src/components/Items/ParcelImage/ParcelImage.js [hunk 9]
```diff
@@ -53,37 +78,30 @@ export default function ParcelImage({ parcel, parcelSize }) {
 
         setImageLoading(true);
 
-        gotchiverseApi.getParcelColorBySizeMap(parcel.parcelId, parcelSize).then(response => {
-            if (mounted) {
-                setImageMap(response);
-            }
-        })
-        .catch(e => console.log(e))
-        .finally(() => setImageLoading(false));
```
Comment (dmutrokot) [inline src/components/Items/ParcelImage/ParcelImage.js:62]: why would you remove `finally` here?

File: src/components/Items/ParcelInstallations/ParcelInstallations.js [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+            />
+        </div>
+    }
+
+    return (
+        <div className={classes.container}>
+            { parcel.installations.map((inst, index) => {
+                const metadata = installationsUtils.getMetadataById(inst.id);
+
+                return <div className={classes.installation} key={index}>
```
Comment (dmutrokot) [inline src/components/Items/ParcelInstallations/ParcelInstallations.js:36]: please wrap into `()`

File: src/contexts/ClientContext.js [hunk 11]
```diff
@@ -281,16 +279,24 @@ const ClientContextProvider = (props) => {
...
-
-            setRealm(commonUtils.basicSort(response, type, dir));
-            setLoadingRealm(false);
-        }).catch((error) => {
-            console.log(error);
-            setRealm([]);
-            setLoadingRealm(false);
-        });
+        thegraph.getRealmByAddress(address)
+            .then(async (res) => {
```
Comment (dmutrokot) [inline src/contexts/ClientContext.js:283]: no need for `async`

File: src/data/citadel.data.js [hunk 12]
```diff
@@ -116,7 +116,8 @@ const DISTRICTS = Object.freeze({
 });
 const COLORS = Object.freeze({
     parcels: {
-        selected: 0xfd9af9,
+        selected: 0xffffff,
+        // selected: 0xfd9af9,
```
Comment (dmutrokot) [inline src/data/citadel.data.js:120]: please remove if not needed

File: src/pages/Client/components/ClientRealmActions.js [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+    const isTransactionProcessing = transactionStatus === 'processing';
+    const isTransactionCompleted = transactionStatus === 'completed';
+
+    useEffect(() => {
+        const accounts = metaState.account;
+        const walletConnected  = accounts.length > 0;
+
+        setIsUserConnected(walletConnected);
+
```
Comment (dmutrokot) [inline src/pages/Client/components/ClientRealmActions.js:28]: no need for whitespace

File: src/pages/Client/components/ClientRealmActions.js [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    const [transactionStatus, setTransactionStatus] = useState('default');
+
+    const { metaState } = useMetamask();
+    const { showSnackbar } = useContext(SnackbarContext);
+
+    const isTransactionProcessing = transactionStatus === 'processing';
+    const isTransactionCompleted = transactionStatus === 'completed';
+
+    useEffect(() => {
+        const accounts = metaState.account;
```
Comment (dmutrokot) [inline src/pages/Client/components/ClientRealmActions.js:24]: is it possible to reuse logic from `LoginContext`?

File: src/pages/Client/components/ClientRealmActions.js [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+        const walletConnected  = accounts.length > 0;
+
+        setIsUserConnected(walletConnected);
+
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [metaState]);
+
+    const onUpgradesFinish = (ids) => {
+        const succesMessage = `successefully finished ${ids.length} upgrades`;
```
Comment (dmutrokot) [inline src/pages/Client/components/ClientRealmActions.js:33]: `successefully` typo

File: src/pages/Client/components/ClientRealmActions.js [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+        const errorMessage = 'upgrades finishing went wrong :(';
```
Comment (dmutrokot) [inline src/pages/Client/components/ClientRealmActions.js:34]: `upgrades finishing` maybe should be vice versa

File: src/pages/Client/components/ClientRealmActions.js [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+            case 'completed':
+                return (
+                    <>
+                        Completed! <CheckIcon size='small' />
+                    </>
+                );
+            case 'processing':
+                return (
+                    <>
+                        Finishing <CircularProgress size={20} color='secondary' sx={{ marginLeft: '8px' }} />
```
Comment (dmutrokot) [inline src/pages/Client/components/ClientRealmActions.js:63]: do we prefer to use inline styles?

File: src/pages/Client/components/ClientRealmActions.js [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+        }
+    }
+
+    return (
+        <div>
+            <Button
+                variant='contained'
+                color={isTransactionCompleted ? 'success' : 'info'}
+                fullWidth
+                sx={{ marginBottom: '8px' }}
```
Comment (dmutrokot) [inline src/pages/Client/components/ClientRealmActions.js:81]: do we prefer to use inline styles?

File: src/pages/Client/components/ClientRealmActions.js [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+                onClick={() => onUpgradesFinish(claimableList)}
+                disabled={!claimableList.length || !isUserConnected || isTransactionProcessing}
+                className={isTransactionCompleted ? classes.buttonCompleted : null}
+            >
+                {renderButtonNode(transactionStatus)}
+            </Button>
+
+            { !isUserConnected && (
+                <Alert severity='error' sx={{ marginBottom: '8px' }}>
```
Comment (dmutrokot) [inline src/pages/Client/components/ClientRealmActions.js:90]: do we prefer to use inline styles?

File: src/pages/Client/routes/ClientRealmList.js [hunk 20]
```diff
@@ -103,117 +136,163 @@ export default function ClientRealmList() {
...
+        const parcelIds = realm.map(parcel => parcel.tokenId);
+
+        setLoadingUpgrades(true);
+
+        Promise.all([
+            getRealmInfo(activeAddress, realm),
+            getRealmUpgradesQueue(parcelIds),
+        ]).then(([realmInfo, realmUpgradesQueue]) => {
+            const modifiedParcels = realm.map((parcel, index) => {
+                const parcelUpgrading = realmUpgradesQueue.find(upgrade => upgrade.parcelId === parcel.tokenId);
```
Comment (dmutrokot) [inline src/pages/Client/routes/ClientRealmList.js:216]: `parcelUpgrading` -> `isParcelUpgrading`

File: src/pages/Client/routes/ClientRealmList.js [hunk 21]
```diff
@@ -103,117 +136,163 @@ export default function ClientRealmList() {
...
+                const parcelInfo = realmInfo.find(info => info.id === parcel.tokenId);
+
+                return {
+                    ...parcel,
+                    channeling: parcelInfo,
+                    nextChannel: parcelInfo.nextChannel,
+                    altarLevel: parcelInfo.installations[0].level,
+                    installations: parcelInfo.installations,
+                    upgrading: parcelUpgrading ? parcelUpgrading : undefined,
```
Comment (dmutrokot) [inline src/pages/Client/routes/ClientRealmList.js:225]: can be simplified to `Boolean(parcelUpgrading)`

File: src/pages/Client/routes/ClientRealmList.js [hunk 22]
```diff
@@ -103,117 +136,163 @@ export default function ClientRealmList() {
...
+                    installations: installations
+                };
+            });
+        })
+    }
+
+    const getRealmUpgradesQueue = (realmIds) => {
+        return installationsApi.getAllUpgradeQueue().then(async res => {
+            const activeUpgrades = res
+                .map((que, i) => ({ ...que, upgradeIndex: i })) // add indexes (needed for onUpgradesFinish function)
```
Comment (dmutrokot) [inline src/pages/Client/routes/ClientRealmList.js:262]: what is `que`?

File: src/pages/Client/routes/ClientRealmList.js [hunk 23]
```diff
@@ -103,117 +136,163 @@ export default function ClientRealmList() {
...
+                        ...upgrade,
+                        timestamp: timestamp,
+                        ready: isUpgradeReady
+                    }
+                });
+
+                setClaimableUpgrades(
+                    upgradesWithTimestamps.filter(que => que.ready).map(que => que.upgradeIndex)
+                );
+                return upgradesWithTimestamps;
```
Comment (dmutrokot) [inline src/pages/Client/routes/ClientRealmList.js:284]: missed whitespace

File: src/pages/Client/routes/ClientRealmList.js [hunk 24]
```diff
@@ -103,117 +136,163 @@ export default function ClientRealmList() {
...
+
+                    return {
+                        ...upgrade,
+                        timestamp: timestamp,
+                        ready: isUpgradeReady
+                    }
+                });
+
+                setClaimableUpgrades(
+                    upgradesWithTimestamps.filter(que => que.ready).map(que => que.upgradeIndex)
```
Comment (dmutrokot) [inline src/pages/Client/routes/ClientRealmList.js:282]: still, what is `que` ?:)

File: src/utils/installationsUtils.js [hunk 25]
```diff
@@ -1,9 +1,27 @@
...
+            spillRate: this.getSpillRateById(id),
+            craftTime: this.getCraftTimeById(id),
+            alchemicaCost: this.getAlchemicaCostById(id),
+            cooldown: this.getCooldownByLevel(this.getLevelById(id))
+        };
+    },
+
     getNameById(id) {
-        return installations[id][15] || '';
+        return installations[id][15];
```
Comment (dmutrokot) [inline src/utils/installationsUtils.js:20]: to avoid potential bugs and unclarity  please create 'enum' for such magic numbers

Comment (dmutrokot) [reply src/utils/installationsUtils.js:20]: applies to all magic numbers in the file related to the installations

**Review by dmutrokot — CHANGES_REQUESTED** [review]

File: src/components/Items/ParcelImage/ParcelImage.js (hunk 9 above)
Comment (dudendy) [reply src/components/Items/ParcelImage/ParcelImage.js:62]: temporary we don't handle case when data not arrived

File: src/data/citadel.data.js (hunk 12 above)
Comment (dudendy) [reply src/data/citadel.data.js:120]: parcel default colors needed for further theme implementation

File: src/pages/Client/components/ClientRealmActions.js (hunk 14 above)
Comment (dudendy) [reply src/pages/Client/components/ClientRealmActions.js:24]: not really, a bit different case
`LoginContext` - app login state
`metaState` - metamask login state

File: src/pages/Client/components/ClientRealmActions.js (hunk 17 above)
Comment (dudendy) [reply src/pages/Client/components/ClientRealmActions.js:63]: in this case I would leave it here to not create separate styles file with 1 line of CSS rule

File: src/pages/Client/components/ClientRealmActions.js (hunk 18 above)
Comment (dudendy) [reply src/pages/Client/components/ClientRealmActions.js:81]: in this case I would leave it here to not create separate styles file with 2 line of CSS rule

File: src/pages/Client/components/ClientRealmActions.js (hunk 19 above)
Comment (dudendy) [reply src/pages/Client/components/ClientRealmActions.js:90]: in this case I would leave it here to not create separate styles file with 3 lines of CSS rule xD

Comment (dudendy) [reply src/pages/Client/components/ClientRealmActions.js:90]: okay okay let it be xd

File: src/pages/Client/routes/ClientRealmList.js (hunk 22 above)
Comment (dudendy) [reply src/pages/Client/routes/ClientRealmList.js:266]: queue

File: src/pages/Client/routes/ClientRealmList.js (hunk 24 above)
Comment (dudendy) [reply src/pages/Client/routes/ClientRealmList.js:286]: queue

File: src/api/installations.api.js [hunk 26]
```diff
@@ -18,7 +18,11 @@ export default {
                 claimed: item.claimed,
                 parcelId: ethersApi.formatBigNumber(item.parcelId),
                 installationId: ethersApi.formatBigNumber(item.installationId),
-            })));
+            })))
+            .catch(e => {
+                console.log(e);
+                return [];
```
Comment (dmutrokot) [inline src/api/installations.api.js:24]: missed whitespace before `return`

File: src/api/installations.api.js [hunk 27]
```diff
@@ -28,19 +32,19 @@ export default {
...
```
Comment (dmutrokot) [inline src/api/installations.api.js:38]: missed whitespace before `return`

File: src/components/ActiveListingButton/ActiveListingButton.js [hunk 28]
```diff
@@ -20,7 +20,8 @@ export default function ActiveListingButton({ item }) {
 
         thegraphApi.getActiveListing(item.erc, item.id, item.type, item.category)
             .then(res => {
-                if(mounted && res) {
+                if (mounted) {
+                    console.log(res)
```
Comment (dmutrokot) [inline src/components/ActiveListingButton/ActiveListingButton.js:24]: please remove

File: src/components/Items/ParcelImage/ParcelImage.js [hunk 29]
```diff
@@ -16,13 +16,13 @@ export default function ParcelImage({ parcel, imageSize }) {
 
         map.forEach((item) => {
             if (!cache.length) {
-                cache[0] = []
+                cache[0] = [];
             }
 
             if (cache[cache.length -1].length < 4) {
```
Comment (dmutrokot) [inline src/components/Items/ParcelImage/ParcelImage.js:22]: missed space after -

File: src/components/Items/ParcelImage/ParcelImage.js [hunk 30]
```diff
@@ -16,13 +16,13 @@ export default function ParcelImage({ parcel, imageSize }) {
...
                 cache[cache.length - 1].push(item)
             } else {
-                cache[cache.length] =[item]
+                cache[cache.length] = [item]
```
Comment (dmutrokot) [inline src/components/Items/ParcelImage/ParcelImage.js:25]: missed semicolon

File: src/data/types.js [hunk 31]
```diff
@@ -32,3 +32,22 @@ export const Erc1155Categories = Object.freeze({
...
+    SpillRate: 6,
+    UpgradeQueueBoost: 7,
+    CraftTime: 8,
+    NextLevelId: 9,
+    Deprecated: 10,
+    AlchemicaCost: 11,
+    HarvestRate: 12,
+    Capacity: 13,
+    Prerequisites: 14,
+    Name: 15,
```
Comment (dmutrokot) [inline src/data/types.js:52]: no comma needed

**Review by dmutrokot — CHANGES_REQUESTED** [review]

**Review by dmutrokot — APPROVED** [review]

**Pull request merged**